text/0000-bom-in-layer-metadata.md

```diff
@@ -0,0 +1,108 @@
+
+Having the `BOM` be a part of the layer content metadata also means that we can pin-point which layer contributed a BOM entry without having to implement extra logic in downstream buildpacks.
+
+- What is the expected outcome?
+
+The lifecycle is able to generate the final `BOM` by combining the `BOM` entries from top level metadata files like `launch.toml` and `build.toml` and also the ones from `<layer-name>.toml` which handling layer caching correctly i.e. if the layer is cached and re-used in the new build - the `BOM` associated with that layer will also be re-used.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The layer content metadata files would be updated to include - 
+
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
+# arbitrary metadata describing the dependency
+```
+
+For each dependency contributed to the app image or build environment, the buildpack layer:
+
+- SHOULD add a bill-of-materials entry to the bom array describing the dependency, where:
+  - name is REQUIRED.
+  - metadata MAY contain additional data describing the dependency.
+  
+The buildpack SHOULD NOT add bom to a layer content metadata file describing the contents that are not contributed by the same layer. `BOM` entries that are not related to a specific layer should go in the buildpack level `BOM` entries in `launch.toml` or `build.toml`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle would be responsible for the generation of the appropriate metadata files and labels given partial `BOM` entries in the `launch.toml`, `build.toml` and `<layer-name>.toml`. The generation of these artifacts given `BOM` entries in `<layer-name>.toml` would be controlled as follows - 
+
+- If the layer has `types.launch` set to `true` then all the `bom` entries will be contributed to the app image.
+- If the layer has `types.build` set to `true` then all the `bom` entries will be contributed to the build environment.
+- If the layer has `types.cache` set to `true` and if the layer is reused in the next build, the respective `bom` entries should be re-populated from the cached layer content metadata file.
+- If the layer has `types.cache` set to `true` and if the layer is reused in the next build, the respective `bom` entries should be re-populated from the cached layer content metadata file.
+- If `types.cache` is set to `false`, but `types.launch` is set to `true` before a rebuild, the layer content metdata file should be restored without the `BOM` section. If the buildpack sets `types.launch` to `true` after the re-build, and leaves an empty layer directory AND an empty BOM section, the BOM should be restored from the previous app image along with the layer. If the buildpack deletes the `<layer-name>.toml` file or fails to set `types.launch` to true, the `BOM` related to that layer should be cleared. If the buildpack leaves a non-empty layer directory or a non-empty bom section, the value of `bom` after the re-build should be used.
```

Comment:
What is the advantage of restoring `<layer>.toml` without a BOM? I would think the information contained there could be useful to the buildpack in deciding whether to re-use the layer.

Reply:
My instinct is that the lifecycle should just trust that whatever BOM is present in the metadata file at the end of the build is the one that the buildpack intended to contribute. But maybe I'm missing something...

Reply:
> What is the advantage of restoring `<layer>.toml` without a BOM? I would think the information contained there could be useful to the buildpack in deciding whether to re-use the layer.

I am a bit cautious about this as that would mean we would have both `metadata` and `bom` to store data to make decisions related to `cache` but maybe I am overthinking here.

Reply:
> My instinct is that the lifecycle should just trust that whatever BOM is present in the metadata file at the end of the build is the one that the buildpack intended to contribute. But maybe I'm missing something...

Hmm - the only reason I added this specific behavior is because I wanted to tie the layer's lifecycle to the BOM.

Reply:
I just think it would be a lot easier to understand :)

Reply:
Updated it, does that look better?

Reply:
This looks great!